Mathlib/AlgebraicGeometry/Noetherian.lean

```diff
@@ -0,0 +1,335 @@
+      IsLocalization.ideal_eq_iInf_comap_map_away hS (I n),
+      iInf_subtype', iInf_subtype']
+  apply iInf_congr
+  intro s
+  congr 1
+  rw [← hSuit s N (hN s)]
+  exact hSuit s n <| Nat.le_trans (hN s) hn
+
+end localizationProps
+
+variable {X : Scheme}
+
+/-- If a scheme `X` has a cover by affine opens whose sections are Noetherian rings,
+then `X` is locally Noetherian. -/
+theorem isLocallyNoetherian_of_affine_cover {ι} {S : ι → X.affineOpens}
+    (hS : (⨆ i, S i : Opens X) = ⊤)
+    (hS' : ∀ i, IsNoetherianRing Γ(X, S i)) : IsLocallyNoetherian X := by
+  refine ⟨fun U => ?_⟩
+  induction U using of_affine_open_cover S hS with
+  | basicOpen U f hN =>
+    have := U.prop.isLocalization_basicOpen f
+    exact IsLocalization.isNoetherianRing (.powers f) Γ(X, X.basicOpen f) hN
+  | openCover U s _ hN =>
+    apply isNoetherianRing_of_away s ‹_›
+    intro ⟨f, hf⟩
+    have : IsNoetherianRing Γ(X, X.basicOpen f) := hN ⟨f, hf⟩
+    have := U.prop.isLocalization_basicOpen f
+    have hEq := IsLocalization.algEquiv (.powers f) (Localization.Away f) Γ(X, X.basicOpen f)
+    exact isNoetherianRing_of_ringEquiv Γ(X, X.basicOpen f) hEq.symm.toRingEquiv
+  | hU => exact hS' _
+
+/-- A scheme is locally Noetherian if and only if it is covered by affine opens whose sections
+are noetherian rings.
+
+See [Har77], Proposition II.3.2. -/
+theorem isLocallyNoetherian_iff_of_iSup_eq_top {ι} {S : ι → X.affineOpens}
+    (hS : (⨆ i, S i : Opens X) = ⊤) :
+    IsLocallyNoetherian X ↔ ∀ i, IsNoetherianRing Γ(X, S i) :=
+  ⟨fun _ i => IsLocallyNoetherian.component_noetherian (S i),
+   isLocallyNoetherian_of_affine_cover hS⟩
+
+open CategoryTheory in
+/-- A version of `isLocallyNoetherian_iff_of_iSup_eq_top` using `Scheme.OpenCover`. -/
+theorem isLocallyNoetherian_iff_of_affine_openCover {𝒰 : Scheme.OpenCover.{v, u} X}
+    [∀ i, IsAffine (𝒰.obj i)] :
+    IsLocallyNoetherian X ↔
+    ∀ (i : 𝒰.J), IsNoetherianRing Γ(𝒰.obj i, ⊤) := by
+  constructor
+  · intro h i
+    let U := Scheme.Hom.opensRange (𝒰.map i)
+    have := h.component_noetherian ⟨U, isAffineOpen_opensRange _⟩
+    apply isNoetherianRing_of_ringEquiv (R := Γ(X, U))
+    apply CategoryTheory.Iso.commRingCatIsoToRingEquiv
+    exact (IsOpenImmersion.ΓIso_top (𝒰.map i)).symm
+  · intro hCNoeth
+    let fS i : X.affineOpens := ⟨Scheme.Hom.opensRange (𝒰.map i), isAffineOpen_opensRange _⟩
+    apply isLocallyNoetherian_of_affine_cover (S := fS)
+    rw [← Scheme.OpenCover.iSup_opensRange 𝒰]
+    intro i
+    apply isNoetherianRing_of_ringEquiv (R := Γ(𝒰.obj i, ⊤))
+    apply CategoryTheory.Iso.commRingCatIsoToRingEquiv
+    exact IsOpenImmersion.ΓIso_top (𝒰.map i)
+
+/-- If `𝒰` is an open cover of a scheme `X`, then `X` is locally noetherian if and only if
+`𝒰.obj i` are all locally noetherian. -/
+theorem isLocallyNoetherian_iff_openCover (𝒰 : Scheme.OpenCover X) :
+    IsLocallyNoetherian X ↔ ∀ (i : 𝒰.J), IsLocallyNoetherian (𝒰.obj i) := by
```

Comment:
Can you separate "if `X -> Y` is an open immersion and `Y` is locally noetherian, then so is `X`" from the => direction to a separate lemma?